source/common/memory/stats.cc

```diff
@@ -2,121 +2,179 @@
-  return value;
-}
-
-uint64_t Stats::totalCurrentlyReserved() {
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("generic.heap_size", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalThreadCacheBytes() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetNumericProperty("tcmalloc.current_total_thread_cache_bytes")
+      .value_or(0);
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("tcmalloc.current_total_thread_cache_bytes",
                                                   &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPageHeapFree() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetNumericProperty("tcmalloc.pageheap_free_bytes").value_or(0);
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("tcmalloc.pageheap_free_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPageHeapUnmapped() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetNumericProperty("tcmalloc.pageheap_unmapped_bytes")
+      .value_or(0);
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("tcmalloc.pageheap_unmapped_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 uint64_t Stats::totalPhysicalBytes() {
+#if defined(TCMALLOC)
+  return tcmalloc::MallocExtension::GetProperties()["generic.physical_memory_used"].value;
+#elif defined(GPERFTOOLS_TCMALLOC)
   size_t value = 0;
   MallocExtension::instance()->GetNumericProperty("generic.total_physical_bytes", &value);
   return value;
+#else
+  return 0;
+#endif
 }
 
 void Stats::dumpStatsToLog() {
+#if defined(TCMALLOC)
+  ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", tcmalloc::MallocExtension::GetStats());
+#elif defined(GPERFTOOLS_TCMALLOC)
   constexpr int buffer_size = 100000;
   auto buffer = std::make_unique<char[]>(buffer_size);
   MallocExtension::instance()->GetStats(buffer.get(), buffer_size);
   ENVOY_LOG_MISC(debug, "TCMalloc stats:\n{}", buffer.get());
+#else
+  return;
+#endif
 }
 
-} // namespace Memory
-} // namespace Envoy
-
-#else
+AllocatorManager::AllocatorManager(
+    Api::Api& api, Envoy::Stats::Scope& scope,
+    const envoy::config::bootstrap::v3::MemoryAllocatorManager& config)
+    : bytes_to_release_(config.bytes_to_release()),
+      memory_release_interval_msec_(std::chrono::milliseconds(
+          PROTOBUF_GET_MS_OR_DEFAULT(config, memory_release_interval, 1000))),
+      allocator_manager_stats_(MemoryAllocatorManagerStats{
+          MEMORY_ALLOCATOR_MANAGER_STATS(POOL_COUNTER_PREFIX(scope, "tcmalloc."))}),
+      api_(api) {
+#if defined(GPERFTOOLS_TCMALLOC)
+  RELEASE_ASSERT((bytes_to_release_ == 0), "Memory releasing is not supported for gperf tcmalloc.");
```

Comment:
I'm confused - probably missing something.  Isn't this created by default in server.cc, and will then release assert by default if we are in this define block?  Shouldn't we gracefully reject config in any case?

Reply:
By default (if user has not explicitly configured memory releasing),  `bytes_to_release_` will be 0 and this release assert will not fail. But it catches the case when user has enabled memory releasing (by specifying how many bytes to release per cycle and the value is > 0) in combination with gperf tcmalloc (which we don't support, we only support tcmalloc (aka "Google tcmalloc"))

Reply:
totally fair but if you load a config the user doesn't suppor the preferred behavior is rejecting config (throwing error / returning absl::Status failure)  not crashing on start up.  Is this not plausible in this case?